tsl/src/compression/recompression.c

```diff
@@ -0,0 +1,1457 @@
+ */
+static void
+save_affected_compressed_rows_using_indexscan(
+	const RowDecompressor *decompressor, const RowCompressor *row_compressor,
+	const Chunk *compressed_chunk, const Chunk *uncompressed_chunk,
+	const CompressTuplesortstateCxt *uncompressed_tuplestorecxt,
+	AffectedSegmentsCxt **affectedsegments, const int nsegmentby_cols)
+{
+	Relation index_rel;
+	IndexScanDesc index_scan;
+	HeapTuple compressed_tuple = NULL;
+	TupleTableSlot *uncompressed_slot = NULL;
+	TupleTableSlot *compressed_slot = NULL;
+	ScanKeyData *segment_by_keys = NULL;
+	ScanKeyData *segment_order_by_keys = NULL;
+
+	int num_segkeys = 0;
+	int num_scankeys = 0;
+	unsigned int num_uncompressedrows = 0;
+	unsigned int uncompressed_row_count = 0;
+
+	bool compressed_tuple_found = false;
+	bool segment_exists = false;
+
+	/* Index scan */
+	index_rel = index_open(row_compressor->index_oid, AccessShareLock);
+	index_scan = index_beginscan(decompressor->in_rel,
+								 index_rel,
+								 GetTransactionSnapshot(),
+								 nsegmentby_cols,
+								 0);
+
+	compressed_slot = table_slot_create(decompressor->in_rel, NULL);
+	uncompressed_slot = MakeTupleTableSlot(decompressor->out_desc, &TTSOpsMinimalTuple);
+
+	/*
+	 * 1. For every uncompressed row do following:
+	 *   A. If fetched compressed tuple (CR) from (iii) exists:
+	 *       - If uncompressed row matching CR goto step1.
+	 *       - If uncompressed row does not match, save total
+	 *       - fetched uncompressed row count and CR in AffectedSegmentsCxt
+	 *   B. Scan the compressed index
+	 *       i. Fetch the compressed tuple based on segmentby values
+	 *       ii. If matching segmentby tuple is found check the orderby values
+	 *       iii. Save the matching segmentby/orderby compressed row (CR)
+	 *       iv. If segmentby values does not match goto stepB
+	 *       v. If none of compressed rows match the segmentby values mark
+	 *          new_segments_found = true and process all new segmentby uncompressed
+	 *          rows.
+	 *   C. Handle all corner cases as part of steps A,B
+	 */
+	while (read_next_uncompressed_tuple(uncompressed_tuplestorecxt->tuplestore, uncompressed_slot))
+	{
+		Bitmapset *null_columns = NULL;
+		ScanKeyData *idx_scankey = NULL;
+		int num_idx_scankeys = 0;
+
+		num_uncompressedrows++;
+		slot_getallattrs(uncompressed_slot);
+
+		if (segment_by_keys)
+			pfree(segment_by_keys);
+		if (segment_order_by_keys)
+			pfree(segment_order_by_keys);
+		/* build scan keys to do lookup into uncompressed chunks */
+		segment_by_keys = build_segment_order_by_scankeys(decompressor,
+														  uncompressed_chunk,
+														  uncompressed_slot,
+														  &null_columns,
+														  &num_segkeys,
+														  SEGMENTBY_KEYS);
+		segment_order_by_keys = build_segment_order_by_scankeys(decompressor,
+																uncompressed_chunk,
+																uncompressed_slot,
+																NULL,
+																&num_scankeys,
+																SEGMENTBY_KEYS | ORDERBY_KEYS);
```

Comment:
Doing this for each uncompressed rows is a lot of work, you can update the values in ScanKey structures directly instead.